subprojects/version-control/src/integTest/groovy/org/gradle/vcs/internal/ThreadSafeVersionControlSystemSpec.groovy

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.vcs.internal
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.vcs.fixtures.GitRepository
+import org.gradle.vcs.git.internal.DefaultGitVersionControlSpec
+import org.junit.Rule
+
+class ThreadSafeVersionControlSystemSpec extends AbstractIntegrationSpec {
+    @Rule GitRepository repo = new GitRepository(temporaryFolder.getTestDirectory())
+
+    def setup() {
+        buildFile << """
+            import ${VersionControlSystemFactory.canonicalName}
+            import ${DefaultGitVersionControlSpec.canonicalName}
+
+            class GitClone extends DefaultTask {
+                private final VersionControlSystemFactory versionControlSystemFactory
+                
+                @Input
+                URI url = project.uri("${repo.url}")
+
+                @OutputDirectory
+                File outputDir = new File(temporaryDir, "target")
+                
+                @javax.inject.Inject
+                GitClone(VersionControlSystemFactory versionControlSystemFactory) {
+                    this.versionControlSystemFactory = versionControlSystemFactory
+                }
+
+                @TaskAction
+                void populate() {
+                    def spec = new DefaultGitVersionControlSpec()
+                    spec.url = url
+                    def system = versionControlSystemFactory.create(spec)
+                    def refs = system.getAvailableVersions(spec)
+                    system.populate(outputDir, refs[0], spec)
+                    assert new File(outputDir, "repo/.git").exists()
+                }
+            }
+            
+            task clone(type: GitClone)
+        """
+    }
+
+    def "can clone"() {
+        def source = repo.workTree.file("source")
+        source.text = "hello world"
+        repo.commit("initial commit", source)
+        expect:
+        succeeds("clone")
```

Comment:
Some ideas for testing this...

- Check that the appropriate lock file has been created in the right place.
- Setup a multi-project build that each try to clone the same repo into the same path.  Coordinate them all by using a BlockingHttpServer and blocking in a `doFirst` for all of the clone tasks.
- Setup a multi-project build as above, but coordinate them through a HTTP git server we start in the test: Maybe we can use/steal this? http://download.eclipse.org/jgit/site/4.8.0.201706111038-r/apidocs/org/eclipse/jgit/http/server/GitServlet.html
- Setup multiple Gradle builds that all try to clone the same repo into the same path.  Coordinate them all with one of the choices above.
- Create multiple commits/tags that distinguish each clone task from one another and have different content in each tag.  Assert that the final working directory only has no dirty/missing/extra files.
- Setup a multi-project build that try to clone the same repo into *different* paths. Coordinate them and assert that they clone in parallel.